### PR TITLE
Librarian gameplay

### PR DESCRIPTION
**Pull request by TheShuEd**
The librarian is too boring from a gameplay point of view. How about adding some influence to her round? Delve into books and study the background of objects in various parts of the station.

image

### CODE REVIEW

**Comment by KingOnBudget** [comment]
Shedud on his way to make sick ass shit again!!

**Comment by TheShuEd** [comment]
I've redesigned the design very seriously. Now he relies more on communicating with people and working with text than on "scanning" with a special object. This is necessary so that the librarian does not look like another kind of scientist.

**Review by AJCM-git — APPROVED** [review]
I like the main concepts, tho i would prefer if it rebranded the librarian as an "archeologist" to fit more with this mechanic, one thing it mentions is that books would thematically cover "ancient", "alien" and "magic" kinds of books, but I think the latter theme would fit better with the Chaplain and whatever mechanic it gets in the future

**Comment by TheShuEd** [comment]
We discussed this with Keron: it would be interesting if the station could receive weak spells similar to "cantrips" from dnd

**Comment by Hanzdegloker** [comment]
Librarian becomes discount station aligned wizard using ancient tomes

and something something chaplain can pray really hard to do cantrips too

**Comment by Fildrance** [comment]
Overall looks really cool, i'd just add few things to think about:
* waiting is no fun, ss14 needs minigames to fill thouse 'wait 45 seconds to scan artefact' or 'wait 30 seconds to decypher book', math minigames or decryption minigames are easy and can be fun most of the time. You should not care a lot about 'cheating' in thouse, the only thing that metters - they should have server side delays so that it will enforce certain time requirement. Its not mandatory for implementation of librarian, but its very nice thing for game pace. thouse 45 sec won't help in creating socialization opportunity - make it minutes OR make them busy, no micro-obsticles, they are annoying as hell
* give 'preview' of book perk early on, like if you have decyphered small part of book - get 3% instead of 30, make it kinda hidden or obscure hint. Giving whole bonus on completion enforces player to either 'expect for real big payoff in the end', give up due to mismatch of work and payoff, make whole working loop really small. Last part is the one i dispise - it makes feel of effort just PUFF - gone.
* Need to think about the way to make it less like 'science but with books'

**Comment by TheShuEd** [comment]
>Need to think about the way to make it less like 'science but with books'

I came to the idea that a librarian should work with the text, and not with mechanics.  The librarian looks for clue texts, uses them to decipher the text, and is rewarded with texts that can be used.  No mechanisms or special mechanics possible.  Perhaps I will actually encrypt recipes with the Caesar cipher

**Comment by deltanedas** [comment]
> Perhaps I will actually encrypt recipes with the Caesar cipher

now realise that anyone can just copy paste it into `caesar` to decrypt

**Comment by TheShuEd** [comment]
that means it's not the one.  There is a shift with a key word to solve it.  And finding this key will be the task

**Comment by TheShuEd** [comment]
well, make the mechanics of encryption and decryption unrelated to the librarian’s gameplay, so that any text can be encrypted to protect data

**Comment by TheShuEd** [comment]
Vigenère cipher

**Comment by deltanedas** [comment]
its still trivial to crack with a 2009 laptop it should just be the server giving random garbage until you find the key, so if you want to pro ultra powergamer crack it without doing anything you have to ddos the server and get swiftly banned

**Comment by TheShuEd** [comment]
that's why I limit decryption to a 30-second delay.

**Comment by TheShuEd** [comment]
> waiting is no fun, ss14 needs minigames to fill thouse 'wait 45 seconds to scan artefact' or 'wait 30 seconds to decypher book', math minigames or decryption minigames are easy and can be fun most of the time. You should not care a lot about 'cheating' in thouse, the only thing that metters - they should have server side delays so that it will enforce certain time requirement. Its not mandatory for implementation of librarian, but its very nice thing for game pace. thouse 45 sec won't help in creating socialization opportunity - make it minutes OR make them busy, no micro-obsticles, they are annoying as hell

I agree with this, but still, for the first iteration of the mechanics, I will just make a delay to make the PR as small as possible

**Comment by Fildrance** [comment]
> > waiting is no fun, ss14 needs minigames to fill thouse 'wait 45 seconds to scan artefact' or 'wait 30 seconds to decypher book', math minigames or decryption minigames are easy and can be fun most of the time. You should not care a lot about 'cheating' in thouse, the only thing that metters - they should have server side delays so that it will enforce certain time requirement. Its not mandatory for implementation of librarian, but its very nice thing for game pace. thouse 45 sec won't help in creating socialization opportunity - make it minutes OR make them busy, no micro-obsticles, they are annoying as hell
> 
> I agree with this, but still, for the first iteration of the mechanics, I will just make a delay to make the PR as small as possible

Like i said - its fine for now, its just a thing to think about.

**Comment by Fildrance** [comment]
> its still trivial to crack with a 2009 laptop it should just be the server giving random garbage until you find the key, so if you want to pro ultra powergamer crack it without doing anything you have to ddos the server and get swiftly banned

I personally think that solving little puzzles like Caesar cipher can be fun. Lots of games use simple math puzzles to keep player busy and not bored. If someone JUST DONT WANT IT and will copy-paste it into some site or cheat otherwise - heck! Yeah, ok, you are great, get your candy. But server will just force you to sit and wait for 25 more seconds due to hard-coded limits. like i said - it should be a way to pass time, not a way to boost your productivity due to cheating. And it should leave opportunity to not solve minigame, u will just wait x2 longer for example (45 sec vs 90 is not looking that great of a difference.)

**Comment by Fildrance** [comment]
So what about the pacing? How much time it will take to get from obtaining the book to getting ability from it? Optimistically. I mean whole game loop for 1 book.
And you said nothing about getting min-perk before fully dechypering book : D what are your ideas about that?

**Comment by TheShuEd** [comment]
> I personally think that solving little puzzles like Caesar cipher can be fun. Lots of games use simple math puzzles to keep player busy and not bored. If someone JUST DONT WANT IT and will copy-paste it into some site or cheat otherwise - heck! Yeah, ok, you are great, get your candy. But server will just force you to sit and wait for 25 more seconds due to hard-coded limits. like i said - it should be a way to pass time, not a way to boost your productivity due to cheating. And it should leave opportunity to not solve minigame, u will just wait x2 longer for example (45 sec vs 90 is not looking that great of a difference.)

I agree, in this case I will reduce the delay to 5 seconds. Its existence is due only to protection against server overload.

**Comment by TheShuEd** [comment]
> So what about the pacing? How much time it will take to get from obtaining the book to getting ability from it? Optimistically. I mean whole game loop for 1 book.
I don't really understand how to calculate the optimal study time, I will be glad for a hint. Considering that we have a finite number of encrypted keywords per round, which can also be repeated from round to round, I think the speed of studying books will gradually increase by the end of the round. The longer the game goes on, the more words the librarian finds, the higher the chance that he will already have researched keys for new books found.
>And you said nothing about getting min-perk before fully dechypering book : D what are your ideas about that?
Can you rephrase that? I didn't quite understand the question

**Comment by Fildrance** [comment]
not optimal, optimistic. By that i mean:
time to take a book to translate + time to find every reference book-dictionary (consider that he will find needed one first try) and finish decypher. What is minimal time to do all the stuff and reap benefit? And how much time will it seems for now to take on average.

Amout mini perk - i meant that if from 0 to 99% book gives nothing, then it will be harder to balance how complex decyphering process is. But if on some gates (15%/33%) you will give some small perk - miniature version of what book is going to give after decypher - then you will have more options and it will be easier to balance numbers while still providing player for motivation. They should know that even if they won't get 'final one' book with passkey, their effort will still give benefits. And as you say that all of stuff is random - doing whole thing can be too difficult in some cases.

**Comment by K-Dynamic** [comment]
We need librarians to decrypt randomised OP chem recipes

**Comment by AJCM-git** [comment]
Fix conflicts and i'll merge
